File: 36-valid-sudoku/valid-sudoku.cpp
class Solution {
public:
    bool valid(vector<vector<char>>& board, int row, int col){
        for(int i=0; i<9; i++){
            if(row != i && col != i && (board[i][col] == board[row][col] || board[row][i] == board[row][col])) return false;
        }
        int x = row < 3 ? 0 : (row < 6 ? 3 : 6),y = col < 3 ? 0 : (col < 6 ? 3 : 6);
        // cout << row << " " << col << " = " << x << " " << y << endl;
        for(int i=0; i<9; i++){
            int r = x+(i/3), c = y+(i%3);
            if(r != row && c != col && board[r][c] == board[row][col]) return false;
        }

        return true;
    }
    bool isValidSudoku(vector<vector<char>>& board) {
        
        for(int i=0; i<9; i++){
            for(int j=0; j<9; j++){
                if(board[i][j] != '.' && !valid(board, i, j)){
                    // cout << board[i][j] << endl;
                    return false;
                } 
            }
        }

        return true;
    }
};

File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int sz = 9;
        int vis[] = new int[9];

        for(int i=0; i<sz; i++){
            Arrays.fill(vis, 0);
            for(int j=0; j<sz; j++){
                if(board[i][j] != '.'){
                    if(vis[board[i][j] - '1'] == 1) return false;
                    vis[board[i][j] - '1'] = 1;
                }
            }
        }

        for(int i=0; i<sz; i++){
            Arrays.fill(vis, 0);
            for(int j=0; j<sz; j++){
                if(board[j][i] != '.'){
                    if(vis[board[j][i] - '1'] == 1) return false;
                    vis[board[j][i] - '1'] = 1;
                }
            }
        }

        for(int i=0; i<27; i+=3){
            int sc = i%9, sr = (i/9) * 3;
            Arrays.fill(vis, 0);
            for(int j=0; j<3; j++){
                for(int k=0; k<3; k++){
                    int r = j + sr, c = k + sc;
                    // System.out.println(r + " " + c);
                    if(board[r][c] != '.'){
                        if(vis[board[r][c] - '1'] == 1) return false;
                        vis[board[r][c] - '1'] = 1;
                    }
                }
            }
        }

        return true;
    }
}